models/workload.go

```diff
@@ -473,23 +485,49 @@ func (workload *Workload) HasIstioSidecar() bool {
 
 // IsGateway return true if the workload is Ingress, Egress or K8s Gateway
 func (workload *Workload) IsGateway() bool {
-	conf := config.Get()
-	if workload.Type == "Deployment" {
-		if labelValue, ok := workload.Labels["operator.istio.io/component"]; ok && (labelValue == "IngressGateways" || labelValue == "EgressGateways") {
-			return true
-		}
-		ingressLabel := conf.GatewayLabel(conf.IstioLabels.IngressGatewayLabel)
-		if labelValue, ok := workload.Labels[ingressLabel[0]]; ok && labelValue == ingressLabel[1] {
-			return true
-		}
-		egressLabel := conf.GatewayLabel(conf.IstioLabels.EgressGatewayLabel)
-		if labelValue, ok := workload.Labels[egressLabel[0]]; ok && labelValue == egressLabel[1] {
-			return true
-		}
-		if _, ok := workload.Labels[conf.IstioLabels.K8sGatewayLabelName]; ok {
-			return true
-		}
+	// There's not consistent labeling for gateways.
+	// In case of using istioctl, you get:
+	// istio: ingressgateway
+	// or
+	// istio: egressgateway
+	//
+	// In case of using helm, you get:
+	// istio: <gateway-name>
+	//
+	// In case of gateway injection you get:
+	// istio: <gateway-name>
+	//
+	// In case of gateway-api you get:
+	// istio.io/gateway-name: gateway
+	//
+	// In case of east/west gateways you get:
+	// istio: eastwestgateway
+	//
+	// We're going to do different checks for all the ways you can label/deploy gateways
+
+	// istioctl
+	if labelValue, ok := workload.Labels["operator.istio.io/component"]; ok && (labelValue == "IngressGateways" || labelValue == "EgressGateways") {
+		return true
 	}
+
+	// There's a lot of unit tests that look specifically for istio: ingressgateway and istio: egressgateway.
+	// These should be covered by istioctl and gateway injection cases but adding checks for these just in case.
+	if labelValue, ok := workload.Labels["istio"]; ok && (labelValue == "ingressgateway" || labelValue == "egressgateway") {
+		return true
+	}
+
+	// Gateway injection. Includes helm because the helm template uses gateway injection.
+	// If the pod injection template is a gateway then it's a gateway.
+	if workload.TemplateAnnotations != nil && workload.TemplateAnnotations["inject.istio.io/templates"] == "gateway" {
+		return true
+	}
+
+	// gateway-api
+	// If this label exists then it's a gateway
+	if _, ok := workload.Labels["istio.io/gateway-name"]; ok {
+		return true
+	}
+
```

Comment:
Should we include here some other labels? When I create the Gateway in Ambient mesh for the istio-ingress namespace I get the missing sidecar. It has another labels: 

![image](https://github.com/user-attachments/assets/ede54def-9c61-4464-a194-0bfbb3a84993)

Maybe the `gateway.istio.io/managed=istio.io-gateway-controller`?

Reply:
This looks like it changed in 1.24 according to the release notes:
> Removed the deprecated istio.io/gateway-name label, please use gateway.networking.k8s.io/gateway-name label instead.

Reply:
> gateway.istio.io/managed

Just using this might be enough. If we use `gateway.networking.k8s.io/gateway-name` I think we'll want to use `gateway.istio.io/managed` as well anyways because there could be other gateway-api implementations in the cluster that aren't istio and we don't want to get those workloads mixed up with the istio gateways.

Reply:
I actually wonder if we're doing this backwards. Shouldn't a `Gateway` be whatever workload is selected by either an Istio `Gateway` resource or a gateway-api `Gateway` resource? We can use that same selector to find the pods/deployments for the gateways rather than having to look at all these different labels.

Reply:
> I actually wonder if we're doing this backwards. Shouldn't a Gateway be whatever workload is selected by either an Istio Gateway resource or a gateway-api Gateway resource? We can use that same selector to find the pods/deployments for the gateways rather than having to look at all these different labels.

I ended up not going with this approach because it would require a lot of refactoring since every workload needs to know if it is a gateway or not due to how the missing sidecar checks work. I wonder if this check in general is somewhat backwards... maybe it should be a positive check like a green check mark if the workload is in the mesh and not a negative check like what it is today that a workload is out of the mesh.

Reply:
I think I'm -1 on that approach.  I think users usually expect us to flag the outlier.  As an aside, aren't all gateway workloads "kind: gateway" ?

Reply:
This was just a shower thought. Not making any changes for this.

>  I think users usually expect us to flag the outlier.

Workloads without a sidecar are only an outlier in namespaces that are part of the mesh. Kiali shows you workloads across your entire cluster. More than likely some parts of cluster is not part of the mesh and you don't want them to be. A good example of this is the Sail Operator. When you deploy the Sail Operator into its own namespace, Kiali complains that it is not part of the mesh. But you don't want the Sail Operator to ever be part of your mesh so this shouldn't be an error. On the other hand, if you have a workload in a namespace that has injection enabled or is part of ambient and that workload isn't part of the mesh, then you definitely want an error/warning.

>  As an aside, aren't all gateway workloads "kind: gateway" ?

Kind of... there's two different `kind: Gateway`s. There's the classic Istio `kind: Gateway` and then there's the Gateway API `kind: Gateway`. When you create a classic Istio `kind: Gateway`, you still have to create the deployment for the Istio proxy yourself. In the Istio `kind: Gateway`, you just specify a selector to reference the gateway Istio proxy pods in the deployment. For Gateway API `kind: Gateway`, the Istio gateway proxy deployment gets created for you when you create the `kind: Gateway` object. From Kiali's perspective, we only care about the actual Istio proxy deployment in both cases since that is what we consider as the "workload" so we don't care about the `kind: Gateway` at all.